pkg/generate/cspc.go

```diff
@@ -181,48 +201,84 @@ func makePools(poolType string, nDevices int, bd map[string][]v1alpha1.BlockDevi
 			})
 		}
 		return &spec, nil
-	} else if poolType == string(cstorv1.PoolMirrored) {
-		if nDevices%2 != 0 {
-			return nil, fmt.Errorf("mirrored pool requires multiples of two block device")
+	case string(cstorv1.PoolMirrored), string(cstorv1.PoolRaidz), string(cstorv1.PoolRaidz2):
+		min := minCount()[poolType]
+		if nDevices%min != 0 {
+			// there must be min number of devices per RaidGroup
+			return nil, fmt.Errorf("number of devices must be a multiple of %d", min)
+		}
+		if min > nDevices {
+			return nil, fmt.Errorf("insufficient blockdevices expected for %s", poolType)
 		}
+		// 1. Start filling in the devices in their RAID-groups per the hostnames
 		for i, host := range hosts {
 			var raidGroups []cstorv1.RaidGroup
 			// add all BDs to a CSPCs CSPI spec
 			bds := bd[host]
 			if len(bds) < nDevices {
 				return nil, fmt.Errorf("not enough eligible blockdevices found on node %s, want %d, found %d", nodes[i], nDevices, len(bds))
 			}
-			index := 0
-			for d := 0; d < nDevices/2; d++ {
-				raids := []cstorv1.CStorPoolInstanceBlockDevice{{BlockDeviceName: bds[index].Name},
-					{BlockDeviceName: bds[index+1].Name}}
+			// 1. sort the BDs by increasing order
+			sort.Slice(bds, func(i, j int) bool {
+				// sort by increasing order
+				return bds[i].Spec.Capacity.Storage < bds[j].Spec.Capacity.Storage
+			})
+			// 2. Check if close to the desired capacity of the pool can be achieved by minimising disk wastage
+			// 3. Suggest the start and end index for the BDs to be used for the raid group
+			maxIndex := len(bds)
+			if maxIndex < nDevices {
+				return nil, fmt.Errorf("not enough eligible blockdevices found on node %s, want %d, found %d", nodes[i], min, maxIndex)
+			}
+			devices := Generate(v1alpha1.BlockDeviceList{Items: bds})
+			for d := 0; d < nDevices/min; d++ {
+				var raids []cstorv1.CStorPoolInstanceBlockDevice
+				tmp, thisRaidGroup, err := devices.Select(minsize, min)
+				// re-assign the head node of the linked-list for next iteration
+				//
+				devices = tmp
```

Comment:
nit:
```suggestion
				devices, thisRaidGroup, err := devices.Select(minsize, min)
				// re-assign the head node of the linked-list for next iteration
				//
				devices = tmp
```

Reply:
This single line changeset will fail the tests, if I do it, it'll create(thanks to the walrus operator `:=`) a `devices` variable in the scope of the for loop and will overshadow the `devices` in the scope just above it in hierarchy and it'll go out of scope, unless I do a `devices = devices` which seems a bit off.at first to most people, that's why the name `tmp`, I won't mind renaming this `tmp` to `d` as per Golang's naming convention, the length of the variable name should be as long as it remains in scope(since this one would be in scope for a very short time).

Reply:
Sorry, I didn't observe at `:=`... My bad :(